File: docs/js/ai/combat.js
import { CONFIG, TIMING } from '../config.js';
import { STATE } from '../state.js';
import { hasLineOfSight, isWall, gridIndex } from '../grid.js';
import { predictPlayerMovement } from './strategy.js';

// Re-export hasLineOfSight for backward compatibility
export { hasLineOfSight };

// Opponent prediction will be passed as parameter to avoid circular import

/**
 * Check if there's a valid beam path to opponent through the maze
 * Uses same logic as actual beam firing
 * @param {Object} player - AI player
 * @param {Object} opponent - Opponent player
 * @returns {{hasPath: boolean, pathLength: number}} Path info
 */
function checkBeamPath(player, opponent) {
  if (!STATE.maze || STATE.maze.length === 0) {
    return { hasPath: false, pathLength: Infinity };
  }

  const startC = Math.floor((player.x - CONFIG.MAZE_OFFSET_X + 1) / CONFIG.CELL_SIZE);
  const startR = Math.floor((player.y + 1) / CONFIG.CELL_SIZE);
  const endC = Math.floor((opponent.x + opponent.size / 2 - CONFIG.MAZE_OFFSET_X) / CONFIG.CELL_SIZE);
  const endR = Math.floor((opponent.y + opponent.size / 2) / CONFIG.CELL_SIZE);

  const start = gridIndex(startC, startR);
  const end = gridIndex(endC, endR);

  if (!start || !end) {
    return { hasPath: false, pathLength: Infinity };
  }

  // Quick BFS to check path existence and length
  const visited = new Set();
  const queue = [{ cell: start, dist: 0 }];
  visited.add(start);

  const directions = [[0, -1, 0], [1, 0, 1], [0, 1, 2], [-1, 0, 3]];

  while (queue.length > 0) {
    const { cell: curr, dist } = queue.shift();

    if (curr === end) {
      return { hasPath: true, pathLength: dist };
    }

    for (const d of directions) {
      const n = gridIndex(curr.c + d[0], curr.r + d[1]);
      if (n && !visited.has(n) && !curr.walls[d[2]] && !n.walls[(d[2] + 2) % 4]) {
        visited.add(n);
        queue.push({ cell: n, dist: dist + 1 });
      }
    }
  }

  return { hasPath: false, pathLength: Infinity };
}

/**
 * Check if opponent is aiming at player (has path and energy to fire)
 * Uses maze pathfinding since beams navigate through corridors
 * @param {Object} player - AI player to check danger for
 * @param {Object} opponent - Opponent who might be firing
 * @returns {{danger: boolean, direction: string|null, urgency: number}} Danger assessment
 */
export function isOpponentAimingAtMe(player, opponent) {
  // Early exit if opponent can't fire
  if (opponent.boostEnergy < 30) {
    return { danger: false, direction: null, urgency: 0 };
  }

  // Check if opponent has a valid beam path to us
  const { hasPath, pathLength } = checkBeamPath(opponent, player);

  if (!hasPath) {
    return { danger: false, direction: null, urgency: 0 };
  }

  // Calculate direction based on relative position
  const playerCenterX = player.x + player.size / 2;
  const playerCenterY = player.y + player.size / 2;
  const oppCenterX = opponent.x + opponent.size / 2;
  const oppCenterY = opponent.y + opponent.size / 2;

  let direction = null;
  const dx = playerCenterX - oppCenterX;
  const dy = playerCenterY - oppCenterY;
  if (Math.abs(dx) > Math.abs(dy)) {
    direction = dx > 0 ? 'right' : 'left';
  } else {
    direction = dy > 0 ? 'down' : 'up';
  }

  // Urgency based on path length (shorter = more urgent)
  // Path of 1-2 = very urgent, 3-4 = urgent, 5+ = less urgent
  const urgency = Math.max(0, 1 - (pathLength - 1) / 6);

  return { danger: true, direction, urgency };
}

/**
 * Get dodge direction to evade incoming fire
 * Wall-aware version checks both perpendicular directions and prefers wall-free path
 * @param {string} threatDirection - Direction the threat is coming from
 * @param {Object} player - Player object with x, y position (optional for wall-aware dodge)
 * @param {boolean} wallAware - Whether to check walls before choosing direction
 * @returns {{dx: number, dy: number}} Dodge movement vector
 */
export function getDodgeDirection(threatDirection, player = null, wallAware = false) {
  let option1, option2;

  // Determine perpendicular dodge directions based on threat
  switch (threatDirection) {
    case 'left':
    case 'right':
      // Threat is horizontal, dodge vertically
      option1 = { dx: 0, dy: -1 }; // up
      option2 = { dx: 0, dy: 1 };  // down
      break;
    case 'up':
    case 'down':
      // Threat is vertical, dodge horizontally
      option1 = { dx: -1, dy: 0 }; // left
      option2 = { dx: 1, dy: 0 };  // right
      break;
    default:
      return { dx: 0, dy: 0 };
  }

  // If not wall-aware or no player position, use random 50/50
  if (!wallAware || !player) {
    return Math.random() < 0.5 ? option1 : option2;
  }

  // Wall-aware dodge: check both directions for walls
  const checkDist = 4; // pixels to check ahead
  const playerCenterX = player.x + player.size / 2;
  const playerCenterY = player.y + player.size / 2;

  const wall1 = isWall(playerCenterX + option1.dx * checkDist, playerCenterY + option1.dy * checkDist);
  const wall2 = isWall(playerCenterX + option2.dx * checkDist, playerCenterY + option2.dy * checkDist);

  // Prefer wall-free direction
  if (wall1 && !wall2) {
    return option2;
  }
  if (!wall1 && wall2) {
    return option1;
  }

  // Both clear or both blocked - random choice
  return Math.random() < 0.5 ? option1 : option2;
}

export function shouldChargeBeam(player, opponent, currentConfig) {
  if (!currentConfig.TACTICAL_CHARGING_ENABLED) {
    return shouldFireBeamBasic(player, opponent, false, null, currentConfig);
  }

  // Don't waste energy firing at shielded opponents
  if (opponent.shieldActive) {
    return false;
  }

  if (player.boostEnergy < (currentConfig.MIN_CHARGE_ENERGY || 65)) {
    return false;
  }

  // Check if there's a valid path through the maze
  const { hasPath, pathLength } = checkBeamPath(player, opponent);

  if (!hasPath) {
    return false;
  }

  // Charge beam is worth it if opponent is glitched (can't dodge well)
  if (opponent.glitchRemaining(STATE.frameCount) > 60) {
    return pathLength <= 6; // Only if reasonably close
  }

  // Charge beam for close targets (path length 1-4)
  return pathLength <= 4;
}

/**
 * Check if AI should fire beam at opponent
 * BEAMS USE MAZE PATHFINDING - they navigate through corridors, not straight lines!
 * @param {Object} player - AI player
 * @param {Object} opponent - Opponent player
 * @param {boolean} useDistanceCheck - Whether to apply distance-based firing probability
 * @param {Object} opponentPrediction - Optional opponent behavior prediction data
 * @param {Object} currentConfig - AI difficulty configuration (optional)
 * @returns {boolean} True if should fire
 */
export function shouldFireBeamBasic(player, opponent, useDistanceCheck = false, opponentPrediction = null, currentConfig = null) {
  // Don't waste energy firing at shielded opponents
  if (opponent.shieldActive) {
    return false;
  }

  const isInsane = currentConfig?.NAME === 'INSANE';

  // Check if there's a valid path through the maze to hit opponent
  const { hasPath, pathLength } = checkBeamPath(player, opponent);

  if (!hasPath) {
    return false; // No path = no hit possible
  }

  // INSANE: Fire if path exists and is reasonably short
  if (isInsane) {
    // Short path (1-3 cells) = always fire
    if (pathLength <= 3) {
      return true;
    }
    // Medium path (4-6 cells) = fire if we have good energy
    if (pathLength <= 6 && player.boostEnergy > 50) {
      return true;
    }
    // Long path (7+ cells) = only fire if energy is high (opponent might dodge)
    if (pathLength <= 10 && player.boostEnergy > 70) {
      return true;
    }
    return false;
  }

  // Non-INSANE difficulties
  // If opponent frequently shields, consider delaying
  if (opponentPrediction && opponentPrediction.shieldProbability > 0.3) {
    if (Math.random() < 0.4) {
      return false;
    }
  }

  // Apply distance-based firing probability based on path length
  if (useDistanceCheck) {
    let fireChance;
    if (pathLength <= 2) {
      fireChance = 1.0;  // Very close: always fire
    } else if (pathLength <= 4) {
      fireChance = 0.8;  // Close: high chance
    } else if (pathLength <= 6) {
      fireChance = 0.5;  // Medium: moderate chance
    } else {
      fireChance = 0.2;  // Far: low chance
    }
    return Math.random() < fireChance;
  }

  // Default: fire if path is reasonable
  return pathLength <= 8;
}

export function shouldDetonateNearbyMines(player, opponent) {
  if (STATE.mines.length === 0) return false;
  let closeMines = STATE.mines.filter(mine => {
    let distPlayer = Math.hypot(mine.x - player.x, mine.y - player.y);
    let distOpp = Math.hypot(mine.x - opponent.x, mine.y - opponent.y);
    return (mine.owner === player.id || mine.owner === -1) &&
      distOpp < 6 && distPlayer > 5;
  });
  return closeMines.length > 0 && player.boostEnergy > 20;
}

/**
 * Find chokepoints in the maze (cells with limited exits)
 * Chokepoints make excellent mine placement locations
 * @param {Object} nearPlayer - Center position to search around
 * @param {number} radius - Search radius in cells (capped at 5 for performance)
 * @returns {Array} Array of chokepoint cells sorted by tactical value
 */
function findChokepoints(nearPlayer, radius) {
  // Cap radius to prevent expensive searches
  radius = Math.min(radius, 5);

  const chokepoints = [];
  const centerC = Math.floor((nearPlayer.x - CONFIG.MAZE_OFFSET_X) / CONFIG.CELL_SIZE);
  const centerR = Math.floor(nearPlayer.y / CONFIG.CELL_SIZE);

  for (let dc = -radius; dc <= radius; dc++) {
    for (let dr = -radius; dr <= radius; dr++) {
      const c = centerC + dc;
      const r = centerR + dr;

      if (c < 0 || c >= CONFIG.COLS || r < 0 || r >= CONFIG.ROWS) continue;

      const cell = STATE.maze[c + r * CONFIG.COLS];
      if (!cell) continue;

      // Count open passages (walls that are false) - unrolled for performance
      let openPassages = 0;
      if (!cell.walls[0]) openPassages++;
      if (!cell.walls[1]) openPassages++;
      if (!cell.walls[2]) openPassages++;
      if (!cell.walls[3]) openPassages++;

      // Chokepoint = cell with only 1-2 open passages (corridor or dead end)
      if (openPassages <= 2 && openPassages >= 1) {
        const distFromCenter = Math.hypot(dc, dr);
        chokepoints.push({
          cell,
          x: CONFIG.MAZE_OFFSET_X + c * CONFIG.CELL_SIZE + CONFIG.CELL_SIZE / 2,
          y: r * CONFIG.CELL_SIZE + CONFIG.CELL_SIZE / 2,
          passages: openPassages,
          distance: distFromCenter,
          // Tactical value: closer corridors are better
          value: (radius - distFromCenter) * (3 - openPassages)
        });

        // Early exit: we only need a few good chokepoints
        if (chokepoints.length >= 5) {
          return chokepoints.sort((a, b) => b.value - a.value);
        }
      }
    }
  }

  return chokepoints.sort((a, b) => b.value - a.value);
}

/**
 * Record mine placement in player's history for density tracking
 * @param {Object} player - AI player
 * @param {number} x - X position of placed mine
 * @param {number} y - Y position of placed mine
 * @param {number} frameCount - Current frame count
 */
export function recordMinePlacement(player, x, y, frameCount) {
  if (!player.minePlacementHistory) {
    player.minePlacementHistory = [];
  }
  player.minePlacementHistory.push({ x, y, frame: frameCount });
  // Keep only last 10 placements
  if (player.minePlacementHistory.length > 10) {
    player.minePlacementHistory.shift();
  }
}

/**
 * Check if placing a mine at position would cause clustering
 * Enhanced version also checks against recent placement history
 * @param {number} x - X position
 * @param {number} y - Y position
 * @param {Object} player - Player object with minePlacementHistory (optional)
 * @param {number} minDistance - Minimum distance between mines (in cells)
 * @param {number} maxNearby - Maximum mines allowed within minDistance
 * @returns {boolean} True if position is too crowded
 */
function isMineAreaCrowded(x, y, player = null, minDistance = 2, maxNearby = 1) {
  const minDistPixels = minDistance * CONFIG.CELL_SIZE;
  let nearbyCount = 0;

  // Check against existing mines
  for (const mine of STATE.mines) {
    const dist = Math.hypot(mine.x - x, mine.y - y);
    if (dist < minDistPixels) {
      nearbyCount++;
      if (nearbyCount > maxNearby) {
        return true;
      }
    }
  }

  // Check against recent placements (avoid re-mining same spots)
  if (player?.minePlacementHistory) {
    for (const placement of player.minePlacementHistory) {
      if (Math.hypot(placement.x - x, placement.y - y) < minDistPixels * 1.5) {
        nearbyCount++;
        if (nearbyCount > maxNearby) {
          return true;
        }
      }
    }
  }

  return false;
}

export function calculateAdvancedMinePositions(player, opponent, currentConfig) {
  if (!currentConfig.ADVANCED_MINING_ENABLED) {
    let randomCell = STATE.maze[Math.floor(Math.random() * STATE.maze.length)];
    return {
      x: CONFIG.MAZE_OFFSET_X + (randomCell.c * CONFIG.CELL_SIZE) + (CONFIG.CELL_SIZE / 2),
      y: (randomCell.r * CONFIG.CELL_SIZE) + (CONFIG.CELL_SIZE / 2)
    };
  }

  const mineStrategy = currentConfig.MINE_STRATEGY || 'BALANCED';
  const checkDensity = currentConfig.MINE_DENSITY_CHECK || false;

  // CHOKEPOINT: Place mines in narrow corridors along opponent's path
  if (mineStrategy === 'AGGRESSIVE' || mineStrategy === 'BALANCED') {
    // Find chokepoints near opponent's predicted path
    const oppGoalX = CONFIG.MAZE_OFFSET_X + (opponent.goalC * CONFIG.CELL_SIZE) + (CONFIG.CELL_SIZE / 2);
    const oppGoalY = (opponent.goalR * CONFIG.CELL_SIZE) + (CONFIG.CELL_SIZE / 2);

    // Midpoint between opponent and their goal
    const midpoint = {
      x: (opponent.x + oppGoalX) / 2,
      y: (opponent.y + oppGoalY) / 2
    };

    const chokepoints = findChokepoints(midpoint, 5);

    if (chokepoints.length > 0 && Math.random() < 0.6) {
      // Filter chokepoints by mine density if enabled
      let validChokepoints = chokepoints;
      if (checkDensity) {
        validChokepoints = chokepoints.filter(cp => !isMineAreaCrowded(cp.x, cp.y, player));
      }

      if (validChokepoints.length > 0) {
        // Pick one of the top 3 valid chokepoints randomly
        const pick = validChokepoints[Math.floor(Math.random() * Math.min(3, validChokepoints.length))];
        return { x: pick.x, y: pick.y };
      }
    }
  }

  // DEFENSIVE: Protect own goal when losing
  if (mineStrategy === 'DEFENSIVE') {
    let ourGoalX = CONFIG.MAZE_OFFSET_X + (player.goalC * CONFIG.CELL_SIZE) + (CONFIG.CELL_SIZE / 2);
    let ourGoalY = (player.goalR * CONFIG.CELL_SIZE) + (CONFIG.CELL_SIZE / 2);

    // Find chokepoints near our goal
    const goalChokepoints = findChokepoints({ x: ourGoalX, y: ourGoalY }, 5);

    if (goalChokepoints.length > 0 && Math.random() < 0.7) {
      const pick = goalChokepoints[Math.floor(Math.random() * Math.min(3, goalChokepoints.length))];
      return { x: pick.x, y: pick.y };
    }

    // Fallback: ring around goal
    let angle = Math.random() * Math.PI * 2;
    let distFromGoal = CONFIG.CELL_SIZE * 2.5 + Math.random() * CONFIG.CELL_SIZE * 0.5;
    return {
      x: ourGoalX + Math.cos(angle) * distFromGoal,
      y: ourGoalY + Math.sin(angle) * distFromGoal
    };
  }

  // AGGRESSIVE: Intercept opponent's path with prediction
  if (mineStrategy === 'AGGRESSIVE') {
    let oppGoalX = CONFIG.MAZE_OFFSET_X + (opponent.goalC * CONFIG.CELL_SIZE) + (CONFIG.CELL_SIZE / 2);
    let oppGoalY = (opponent.goalR * CONFIG.CELL_SIZE) + (CONFIG.CELL_SIZE / 2);
    let dirToGoal = { x: oppGoalX - opponent.x, y: oppGoalY - opponent.y };
    let dist = Math.hypot(dirToGoal.x, dirToGoal.y);
    if (dist > 0.1) {
      // Place further ahead to catch them
      let interceptDistance = CONFIG.CELL_SIZE * (3 + Math.random() * 3);
      return {
        x: opponent.x + (dirToGoal.x / dist) * interceptDistance,
        y: opponent.y + (dirToGoal.y / dist) * interceptDistance
      };
    }
  }

  // BALANCED: Mix of strategies (non-recursive)
  const roll = Math.random();
  if (roll < 0.5) {
    // AGGRESSIVE inline: Intercept opponent's path
    let oppGoalX = CONFIG.MAZE_OFFSET_X + (opponent.goalC * CONFIG.CELL_SIZE) + (CONFIG.CELL_SIZE / 2);
    let oppGoalY = (opponent.goalR * CONFIG.CELL_SIZE) + (CONFIG.CELL_SIZE / 2);
    let dirToGoal = { x: oppGoalX - opponent.x, y: oppGoalY - opponent.y };
    let dist = Math.hypot(dirToGoal.x, dirToGoal.y);
    if (dist > 0.1) {
      let interceptDistance = CONFIG.CELL_SIZE * (3 + Math.random() * 3);
      return {
        x: opponent.x + (dirToGoal.x / dist) * interceptDistance,
        y: opponent.y + (dirToGoal.y / dist) * interceptDistance
      };
    }
  } else if (roll < 0.8) {
    // Find a chokepoint at midpoint
    const midX = (player.x + opponent.x) / 2;
    const midY = (player.y + opponent.y) / 2;
    const chokepoints = findChokepoints({ x: midX, y: midY }, 5);
    if (chokepoints.length > 0) {
      const pick = chokepoints[Math.floor(Math.random() * Math.min(3, chokepoints.length))];
      return { x: pick.x, y: pick.y };
    }
  }

  // DEFENSIVE fallback: ring around our goal
  let ourGoalX = CONFIG.MAZE_OFFSET_X + (player.goalC * CONFIG.CELL_SIZE) + (CONFIG.CELL_SIZE / 2);
  let ourGoalY = (player.goalR * CONFIG.CELL_SIZE) + (CONFIG.CELL_SIZE / 2);
  let angle = Math.random() * Math.PI * 2;
  let distFromGoal = CONFIG.CELL_SIZE * 2.5 + Math.random() * CONFIG.CELL_SIZE * 0.5;
  return {
    x: ourGoalX + Math.cos(angle) * distFromGoal,
    y: ourGoalY + Math.sin(angle) * distFromGoal
  };
}

File: pi/ai/combat.py
# pi/ai/combat.py
# Combat decision helpers ported from docs/js/ai/combat.js

import math
import random
from collections import deque

from config import COLS, ROWS, CELL_SIZE, MAZE_OFFSET_X
from grid import grid_index, is_wall


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

_beam_visited = bytearray(COLS * ROWS)
_beam_dist = [0] * (COLS * ROWS)

def _check_beam_path(maze: list, player, opponent) -> dict:
    """
    Fast BFS check for a valid beam path between player and opponent cells.
    Returns {'has_path': bool, 'path_length': int}.
    """
    if not maze:
        return {'has_path': False, 'path_length': float('inf')}

    sc = int((player.x - MAZE_OFFSET_X + 1) // CELL_SIZE)
    sr = int((player.y + 1) // CELL_SIZE)
    size = getattr(opponent, 'size', 2.0)
    ec = int((opponent.x + size / 2 - MAZE_OFFSET_X) // CELL_SIZE)
    er = int((opponent.y + size / 2) // CELL_SIZE)

    sc = max(0, min(sc, COLS - 1))
    sr = max(0, min(sr, ROWS - 1))
    ec = max(0, min(ec, COLS - 1))
    er = max(0, min(er, ROWS - 1))

    si = sc + sr * COLS
    ei = ec + er * COLS

    if si == ei:
        return {'has_path': True, 'path_length': 0}

    visited = _beam_visited
    dist = _beam_dist
    total = COLS * ROWS
    for i in range(total):
        visited[i] = 0

    visited[si] = 1
    dist[si] = 0
    queue = deque()
    queue.append(si)

    directions = ((0, -1, 0), (1, 0, 1), (0, 1, 2), (-1, 0, 3))

    while queue:
        ci = queue.popleft()
        if ci == ei:
            return {'has_path': True, 'path_length': dist[ci]}

        cell = maze[ci]
        cc, cr = cell.c, cell.r
        d = dist[ci] + 1

        for dc, dr, wi in directions:
            nc, nr = cc + dc, cr + dr
            if nc < 0 or nc >= COLS or nr < 0 or nr >= ROWS:
                continue
            ni = nc + nr * COLS
            if visited[ni]:
                continue
            if cell.walls[wi]:
                continue
            if maze[ni].walls[(wi + 2) & 3]:
                continue
            visited[ni] = 1
            dist[ni] = d
            queue.append(ni)

    return {'has_path': False, 'path_length': float('inf')}


def _find_chokepoints(maze: list, near_pos: dict, radius: int) -> list:
    """
    Find cells with limited exits (chokepoints) near near_pos.
    Returns list of dicts sorted by tactical value.
    """
    radius = min(radius, 5)
    chokepoints = []
    center_c = int((near_pos['x'] - MAZE_OFFSET_X) // CELL_SIZE)
    center_r = int(near_pos['y'] // CELL_SIZE)

    for dc in range(-radius, radius + 1):
        for dr in range(-radius, radius + 1):
            c = center_c + dc
            r = center_r + dr
            if c < 0 or c >= COLS or r < 0 or r >= ROWS:
                continue
            cell = maze[c + r * COLS]
            if cell is None:
                continue

            open_passages = sum(1 for w in cell.walls if not w)

            if 1 <= open_passages <= 2:
                dist_from_center = math.hypot(dc, dr)
                cp = {
                    'cell': cell,
                    'x': MAZE_OFFSET_X + c * CELL_SIZE + CELL_SIZE / 2,
                    'y': r * CELL_SIZE + CELL_SIZE / 2,
                    'passages': open_passages,
                    'distance': dist_from_center,
                    'value': (radius - dist_from_center) * (3 - open_passages),
                }
                chokepoints.append(cp)
                if len(chokepoints) >= 5:
                    chokepoints.sort(key=lambda cp_: cp_['value'], reverse=True)
                    return chokepoints

    chokepoints.sort(key=lambda cp_: cp_['value'], reverse=True)
    return chokepoints


def _is_mine_area_crowded(maze_mines: list, x: float, y: float,
                          player=None, min_distance: int = 2, max_nearby: int = 1) -> bool:
    """Check if placing a mine at (x, y) would cause clustering."""
    min_dist_px = min_distance * CELL_SIZE
    nearby = 0

    for mine in maze_mines:
        if math.hypot(mine['x'] - x, mine['y'] - y) < min_dist_px:
            nearby += 1
            if nearby > max_nearby:
                return True

    if player is not None:
        history = getattr(player, 'mine_placement_history', [])
        for placement in history:
            if math.hypot(placement['x'] - x, placement['y'] - y) < min_dist_px * 1.5:
                nearby += 1
                if nearby > max_nearby:
                    return True

    return False


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def is_opponent_aiming_at_me(maze: list, player, opponent) -> dict:
    """
    Check if opponent has a beam path to the player.
    Returns {'danger': bool, 'direction': str|None, 'urgency': float}.
    Mirrors JS isOpponentAimingAtMe().
    """
    if opponent.boost_energy < 30:
        return {'danger': False, 'direction': None, 'urgency': 0}

    result = _check_beam_path(maze, opponent, player)
    if not result['has_path']:
        return {'danger': False, 'direction': None, 'urgency': 0}

    size = getattr(player, 'size', 2.0)
    opp_size = getattr(opponent, 'size', 2.0)
    p_cx = player.x + size / 2
    p_cy = player.y + size / 2
    o_cx = opponent.x + opp_size / 2
    o_cy = opponent.y + opp_size / 2

    dx = p_cx - o_cx
    dy = p_cy - o_cy
    if abs(dx) > abs(dy):
        direction = 'right' if dx > 0 else 'left'
    else:
        direction = 'down' if dy > 0 else 'up'

    path_len = result['path_length']
    urgency = max(0.0, 1.0 - (path_len - 1) / 6.0)

    return {'danger': True, 'direction': direction, 'urgency': urgency}


def get_dodge_direction(threat_direction: str, player=None, wall_aware: bool = False) -> dict:
    """
    Return {dx, dy} perpendicular dodge vector for the given threat direction.
    Mirrors JS getDodgeDirection().
    """
    if threat_direction in ('left', 'right'):
        option1 = {'dx': 0, 'dy': -1}
        option2 = {'dx': 0, 'dy': 1}
    elif threat_direction in ('up', 'down'):
        option1 = {'dx': -1, 'dy': 0}
        option2 = {'dx': 1, 'dy': 0}
    else:
        return {'dx': 0, 'dy': 0}

    if not wall_aware or player is None:
        return option1 if random.random() < 0.5 else option2

    check_dist = 4
    size = getattr(player, 'size', 2.0)
    cx = player.x + size / 2
    cy = player.y + size / 2

    # We need maze reference — skip wall check when not available
    try:
        from state import get_state  # type: ignore
        maze = get_state().maze
        wall1 = is_wall(maze, cx + option1['dx'] * check_dist, cy + option1['dy'] * check_dist)
        wall2 = is_wall(maze, cx + option2['dx'] * check_dist, cy + option2['dy'] * check_dist)
        if wall1 and not wall2:
            return option2
        if not wall1 and wall2:
            return option1
    except Exception:
        pass

    return option1 if random.random() < 0.5 else option2


def should_charge_beam(maze: list, player, opponent, current_config: dict) -> bool:
    """
    Return True if AI should charge beam at opponent.
    Mirrors JS shouldChargeBeam().
    """
    if not current_config.get('TACTICAL_CHARGING_ENABLED'):
        return should_fire_beam_basic(maze, player, opponent, False, None, current_config)

    if getattr(opponent, 'shield_active', False):
        return False

    if player.boost_energy < current_config.get('MIN_CHARGE_ENERGY', 65):
        return False

    result = _check_beam_path(maze, player, opponent)
    if not result['has_path']:
        return False

    path_len = result['path_length']

    # Check if opponent is glitched
    try:
        from state import get_state  # type: ignore
        frame = get_state().frame_count
        if opponent.glitch_remaining(frame) > 60:
            return path_len <= 6
    except Exception:
        pass

    return path_len <= 4


def should_fire_beam_basic(maze: list, player, opponent,
                            use_distance_check: bool = False,
                            opponent_prediction=None,
                            current_config: dict = None) -> bool:
    """
    Return True if AI should fire a basic beam at opponent.
    Mirrors JS shouldFireBeamBasic().
    """
    if getattr(opponent, 'shield_active', False):
        return False

    is_insane = (current_config or {}).get('NAME') == 'INSANE'
    result = _check_beam_path(maze, player, opponent)

    if not result['has_path']:
        return False

    path_len = result['path_length']

    if is_insane:
        if path_len <= 3:
            return True
        if path_len <= 6 and player.boost_energy > 50:
            return True
        if path_len <= 10 and player.boost_energy > 70:
            return True
        return False

    # Non-INSANE
    if opponent_prediction and opponent_prediction.get('shieldProbability', 0) > 0.3:
        if random.random() < 0.4:
            return False

    if use_distance_check:
        if path_len <= 2:
            fire_chance = 1.0
        elif path_len <= 4:
            fire_chance = 0.8
        elif path_len <= 6:
            fire_chance = 0.5
        else:
            fire_chance = 0.2
        return random.random() < fire_chance

    return path_len <= 8


def should_detonate_nearby_mines(mines: list, player, opponent) -> bool:
    """
    Return True if AI should detonate mines close to the opponent.
    Mirrors JS shouldDetonateNearbyMines().
    """
    if not mines:
        return False

    for mine in mines:
        owner = mine.get('owner', -1)
        if owner not in (player.id, -1):
            continue
        dist_opp = math.hypot(mine['x'] - opponent.x, mine['y'] - opponent.y)
        dist_self = math.hypot(mine['x'] - player.x, mine['y'] - player.y)
        if dist_opp < 6 and dist_self > 5:
            return player.boost_energy > 20

    return False


def calculate_mine_position(maze: list, player, opponent, current_config: dict) -> dict:
    """
    Calculate a strategic mine placement position.
    Returns {'x': float, 'y': float}.
    Mirrors JS calculateAdvancedMinePositions().
    """
    if not current_config.get('ADVANCED_MINING_ENABLED'):
        if maze:
            cell = maze[int(random.random() * len(maze))]
            return {
                'x': MAZE_OFFSET_X + cell.c * CELL_SIZE + CELL_SIZE / 2,
                'y': cell.r * CELL_SIZE + CELL_SIZE / 2,
            }
        return {'x': player.x, 'y': player.y}

    mine_strategy = current_config.get('MINE_STRATEGY', 'BALANCED')
    check_density = current_config.get('MINE_DENSITY_CHECK', False)
    mines = []
    try:
        from state import get_state  # type: ignore
        mines = get_state().mines
    except Exception:
        pass

    opp_goal_x = MAZE_OFFSET_X + opponent.goal_c * CELL_SIZE + CELL_SIZE / 2
    opp_goal_y = opponent.goal_r * CELL_SIZE + CELL_SIZE / 2

    if mine_strategy in ('AGGRESSIVE', 'BALANCED'):
        midpoint = {
            'x': (opponent.x + opp_goal_x) / 2,
            'y': (opponent.y + opp_goal_y) / 2,
        }
        chokepoints = _find_chokepoints(maze, midpoint, 5)

        if chokepoints and random.random() < 0.6:
            valid = chokepoints
            if check_density:
                valid = [cp for cp in chokepoints
                         if not _is_mine_area_crowded(mines, cp['x'], cp['y'], player)]
            if valid:
                pick = valid[int(random.random() * min(3, len(valid)))]
                return {'x': pick['x'], 'y': pick['y']}

    if mine_strategy == 'DEFENSIVE':
        our_goal_x = MAZE_OFFSET_X + player.goal_c * CELL_SIZE + CELL_SIZE / 2
        our_goal_y = player.goal_r * CELL_SIZE + CELL_SIZE / 2
        goal_cps = _find_chokepoints(maze, {'x': our_goal_x, 'y': our_goal_y}, 5)
        if goal_cps and random.random() < 0.7:
            pick = goal_cps[int(random.random() * min(3, len(goal_cps)))]
            return {'x': pick['x'], 'y': pick['y']}
        angle = random.random() * math.pi * 2
        dist_from_goal = CELL_SIZE * 2.5 + random.random() * CELL_SIZE * 0.5
        return {
            'x': our_goal_x + math.cos(angle) * dist_from_goal,
            'y': our_goal_y + math.sin(angle) * dist_from_goal,
        }

    if mine_strategy == 'AGGRESSIVE':
        dir_x = opp_goal_x - opponent.x
        dir_y = opp_goal_y - opponent.y
        dist = math.hypot(dir_x, dir_y)
        if dist > 0.1:
            intercept_dist = CELL_SIZE * (3 + random.random() * 3)
            return {
                'x': opponent.x + (dir_x / dist) * intercept_dist,
                'y': opponent.y + (dir_y / dist) * intercept_dist,
            }

    # BALANCED fallback
    roll = random.random()
    if roll < 0.5:
        dir_x = opp_goal_x - opponent.x
        dir_y = opp_goal_y - opponent.y
        dist = math.hypot(dir_x, dir_y)
        if dist > 0.1:
            intercept_dist = CELL_SIZE * (3 + random.random() * 3)
            return {
                'x': opponent.x + (dir_x / dist) * intercept_dist,
                'y': opponent.y + (dir_y / dist) * intercept_dist,
            }
    elif roll < 0.8:
        mid_x = (player.x + opponent.x) / 2
        mid_y = (player.y + opponent.y) / 2
        cps = _find_chokepoints(maze, {'x': mid_x, 'y': mid_y}, 5)
        if cps:
            pick = cps[int(random.random() * min(3, len(cps)))]
            return {'x': pick['x'], 'y': pick['y']}

    # DEFENSIVE fallback
    our_goal_x = MAZE_OFFSET_X + player.goal_c * CELL_SIZE + CELL_SIZE / 2
    our_goal_y = player.goal_r * CELL_SIZE + CELL_SIZE / 2
    angle = random.random() * math.pi * 2
    dist_from_goal = CELL_SIZE * 2.5 + random.random() * CELL_SIZE * 0.5
    return {
        'x': our_goal_x + math.cos(angle) * dist_from_goal,
        'y': our_goal_y + math.sin(angle) * dist_from_goal,
    }


def record_mine_placement(player, x: float, y: float, frame_count: int) -> None:
    """Record mine placement in player history for density tracking."""
    if not hasattr(player, 'mine_placement_history') or player.mine_placement_history is None:
        player.mine_placement_history = []
    player.mine_placement_history.append({'x': x, 'y': y, 'frame': frame_count})
    if len(player.mine_placement_history) > 10:
        player.mine_placement_history.pop(0)
